src/metabase/lib/schema/ref.cljc

```diff
@@ -15,44 +16,54 @@
    [:map
     [:temporal-unit {:optional true} ::temporal-bucketing/unit]]])
 
+(mr/def ::base-type
+  [:fn #(isa? % :type/*)])
+
+(mr/def ::field.literal.options
+  [:and
+   ::field.options
+   [:map
+    [:base-type ::base-type]]])
+
 ;;; `:field` clause
+(mr/def ::field.literal
+  [:tuple
+   [:= :field]
+   ::field.literal.options
+   ::common/non-blank-string])
+
+(mr/def ::field.id
+  [:tuple
+   [:= :field]
+   ::field.options ; TODO -- we should make `:base-type` required here too
+   ::id/field])
+
 (mr/def ::field
-  [:and
-   [:catn
-    [:clause [:= :field]]
-    [:id-or-name [:altn
-                  [:id ::id/field]
-                  [:name ::common/non-blank-string]]]
-    [:options ::field.options]]
-   [:fn (fn [[_field id-or-name options]]
-          (or (integer? id-or-name)
-              (and (string? id-or-name)
-                   (isa? (:base-type options) :type/*))))]])
-
-;; this is a placeholder that will be resolved later once we have metadata and `append` it to the query. Used to
-;; implement [[metabase.lib.field/field]] so you don't have to pass metadata to it directly.
-(mr/def ::field.unresolved
-  [:catn
-   [:clause [:= :field/unresolved]]
-   [:info [:fn map?]]])
+  [:multi {:dispatch (fn [[_field _opts id-or-name]]
```

Comment:
I was experimenting trying to improve the error messages for `:field` clauses. I'm not convinced I really improved the situation all that much.